File: 21-graphs3/5-strongly-connected-components/main.py
# Strongly Connected Components - Kosaraju's Algorithm - Python Example

from collections import defaultdict

def dfs(graph, v, visited, stack=None):
    visited.add(v)
    for nbr in graph[v]:
        if nbr not in visited:
            dfs(graph, nbr, visited, stack)
    if stack is not None:
        stack.append(v)

def reverse_graph(graph):
    rev = defaultdict(list)
    for u in graph:
        for v in graph[u]:
            rev[v].append(u)
    return rev

def kosaraju_scc(graph):
    visited = set()
    stack = []

    for vertex in graph:
        if vertex not in visited:
            dfs(graph, vertex, visited, stack)

    rev = reverse_graph(graph)
    visited.clear()
    sccs = []

    while stack:
        v = stack.pop()
        if v not in visited:
            comp = []
            dfs(rev, v, visited, comp)
            sccs.append(comp)

    return sccs

graph = {
    0: [1],
    1: [2],
    2: [0, 3],
    3: [4],
    4: []
}

components = kosaraju_scc(graph)
print("Strongly connected components:", components)

File: 21-graphs3/5-strongly-connected-components/main.java
// Strongly Connected Components - Kosaraju's Algorithm - Java Example

import java.util.*;

class KosarajuSCC {

    public static void dfs(Map<Integer, List<Integer>> graph, int v, Set<Integer> visited, Stack<Integer> stack) {
        visited.add(v);
        for (int nbr : graph.getOrDefault(v, Collections.emptyList())) {
            if (!visited.contains(nbr)) {
                dfs(graph, nbr, visited, stack);
            }
        }
        stack.push(v);
    }

    public static void dfsCollect(Map<Integer, List<Integer>> graph, int v, Set<Integer> visited, List<Integer> component) {
        visited.add(v);
        component.add(v);
        for (int nbr : graph.getOrDefault(v, Collections.emptyList())) {
            if (!visited.contains(nbr)) {
                dfsCollect(graph, nbr, visited, component);
            }
        }
    }

    public static Map<Integer, List<Integer>> reverseGraph(Map<Integer, List<Integer>> graph) {
        Map<Integer, List<Integer>> rev = new HashMap<>();
        for (int u : graph.keySet()) {
            for (int v : graph.get(u)) {
                rev.computeIfAbsent(v, k -> new ArrayList<>()).add(u);
            }
        }
        return rev;
    }

    public static List<List<Integer>> kosarajuSCC(Map<Integer, List<Integer>> graph) {
        Set<Integer> visited = new HashSet<>();
        Stack<Integer> stack = new Stack<>();

        for (int v : graph.keySet()) {
            if (!visited.contains(v)) dfs(graph, v, visited, stack);
        }

        Map<Integer, List<Integer>> rev = reverseGraph(graph);
        visited.clear();
        List<List<Integer>> sccs = new ArrayList<>();

        while (!stack.isEmpty()) {
            int v = stack.pop();
            if (!visited.contains(v)) {
                List<Integer> component = new ArrayList<>();
                dfsCollect(rev, v, visited, component);
                sccs.add(component);
            }
        }
        return sccs;
    }

    public static void main(String[] args) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        graph.put(0, Arrays.asList(1));
        graph.put(1, Arrays.asList(2));
        graph.put(2, Arrays.asList(0, 3));
        graph.put(3, Arrays.asList(4));
        graph.put(4, Collections.emptyList());

        List<List<Integer>> sccs = kosarajuSCC(graph);
        System.out.println("Strongly connected components:");
        for (List<Integer> component : sccs) {
            System.out.println(component);
        }
    }
}
